### PR TITLE
Coverage

### PR DESCRIPTION
**Pull request by altendky**
Draft for:
- [x] Make it work
- [ ] Self review

### CODE REVIEW

**Comment by altendky** [comment]
What with the recent [Codecov incident](https://about.codecov.io/security-update/), I decided to keep this in-build for now.  No reporting to external services.  Though honestly, this still adds downloading and running of coverage and diff-cover so...  But anyways.  It checks project coverage (presently requiring 35%, we have 37% or so) and diff coverage (presently requiring 100%).  I kind of feel like we might want to get the project coverage up a lot higher before actually mandating 100% diff coverage.  Forcing small changes to implement basic testing for the code around them is not great.  I think it makes a bit more sense to just make a big push to add tests everywhere, along with some amount of refactor for testability where needed.  We could make the coverage job pass regardless of coverage level for now so it is there for feedback as we do dedicated test addition PRs.  Presently it is not generating HTML reports but in the build log we do get the report below.

```
Name                                       Stmts   Miss    Cover   Missing
--------------------------------------------------------------------------
src/plotman/__init__.py                        0      0   100.0%
src/plotman/__main__.py                        2      2     0.0%   3-10
src/plotman/_tests/__init__.py                 0      0   100.0%
src/plotman/_tests/archive_test.py             7      0   100.0%
src/plotman/_tests/configuration_test.py      36      0   100.0%
src/plotman/_tests/job_test.py                32      0   100.0%
src/plotman/_tests/manager_test.py            38      4    89.5%   58-61
src/plotman/_tests/plot_util_test.py          32      0   100.0%
src/plotman/_tests/reporting_test.py          23      0   100.0%
src/plotman/_tests/resources/__init__.py       0      0   100.0%
src/plotman/analyzer.py                       81     81     0.0%   1-136
src/plotman/archive.py                       105     71    32.4%   22-53, 67, 69, 73, 77, 79, 88-100, 114-123, 130-177
src/plotman/configuration.py                  66      1    98.5%   16
src/plotman/interactive.py                   209    209     0.0%   1-349
src/plotman/job.py                           202    129    36.1%   20, 24, 27-28, 38-49, 83-98, 103-160, 189, 196, 202, 205-253, 257, 260, 279, 282-291, 295-305, 308-309, 312, 315, 318-323, 326-327, 330, 334-338, 346-347
src/plotman/manager.py                        88     43    51.1%   51, 71-138, 141-145
src/plotman/plot_util.py                      52      9    82.7%   9-10, 25, 32-35, 57-58
src/plotman/plotman.py                       156    156     0.0%   1-248
src/plotman/reporting.py                     130     84    35.4%   11-14, 41, 67-125, 129-145, 148-171, 174-186, 194-200
src/plotman/resources/__init__.py              0      0   100.0%
--------------------------------------------------------------------------
TOTAL                                       1259    789    37.3%
```

Thoughts?

**Comment by ericaltendorf** [comment]
I approve of code coverage metrics but know nothing about this system or
how to configure it, so I would only be able to rubber stamp approval.

**Comment by altendky** [comment]
Yeah, presumably something is messed up.  \_Every\_ setup is unique because...  of stuff.  And yeah, that's bad.  But I think the question here is how we want it to act.  I would suggest it just be a monitoring tool until now.  So, neither project coverage nor diff coverage would fail the build.  It's just painful to require that when there's so much coverage to catch up on.  Sound ok?

**Comment by ericaltendorf** [comment]
Yes, I think monitoring is the right call right now.

Is there a template for creating PRs as well?  I couldn't find it, but it
seems like it would be useful to provide some guidance to first time PR
submitters, and we could put some notes in there about testing expectations.

**Comment by altendky** [comment]
https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository

https://github.com/twisted/twisted/blob/63649469c1fe46d8a713e8034239ac3cc0498ea7/.github/PULL_REQUEST_TEMPLATE

https://github.com/ericaltendorf/plotman/issues/73

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -159,12 +168,116 @@ jobs:
...
+        path: all_coverage_report/*
+
   all:
     name: All
     runs-on: ubuntu-latest
     needs:
       - build
       - test
+      # TODO: make this required when we have a better testing situation
+      # - coverage
```
Comment (altendky) [inline .github/workflows/ci.yml:280]: Our GitHub settings only require the All job to pass and this makes that not depend on the coverage job.  So, the coverage job can fail without blocking merging etc but at least the coverage job will still be red for anything below 100% coverage on the diff.

**Pull request merged**